### PR TITLE
feat(ticket#285):generate monthly report as PDF

### PR DESCRIPTION
**Pull request by ghulamhafeez**
## Implementation
- Added cron job
- Create api for generate PDF

### Resources
image

### CODE REVIEW

File: src/app/(app)/attendance-report/components/storePdf.tsx
Comment (muhammad-bilal-ali-khokhar) [inline src/app/(app)/attendance-report/components/storePdf.tsx:1]: I think this file code not using any where you can remove this file

File: src/app/(app)/attendance-report/components/AttendanceReport.tsx
Comment (muhammad-bilal-ali-khokhar) [inline src/app/(app)/attendance-report/components/AttendanceReport.tsx:1]: rename `MonthlyAttendanceReport`

File: src/app/(app)/attendance-report/components/generatePdf.tsx
Comment (muhammad-bilal-ali-khokhar) [inline src/app/(app)/attendance-report/components/generatePdf.tsx:1]: rename `MonthlyReportPdf`

File: tsconfig.json [hunk 1]
```diff
@@ -27,6 +27,6 @@
     }
    
   },
-  "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx", ".next/types/**/*.ts"],
+  "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx", ".next/types/**/*.ts", "src/app/api/cron/notifications/attendance-report/route.tsx", "src/app/(app)/attendance-report/components/MyDocument.js"],
```
Comment (Mubashar1009) [inline tsconfig.json:30]: why we need here this ?

File: src/app/api/cron/notifications/monthly-attendance-report/route.tsx [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    // eslint-disable-next-line react/react-in-jsx-scope
+    const stream = await ReactPDF.renderToStream(
+      // eslint-disable-next-line react/react-in-jsx-scope
+      <MyDocument data={attendanceReport} />
+    );
+    const pdfBuffer = await streamToBuffer(stream);
+
+    await supabase.storage
+      .from(bucketName)
+      .upload(fileName, pdfBuffer, { upsert: true });
```
Comment (Mubashar1009) [inline src/app/api/cron/notifications/monthly-attendance-report/route.tsx:42]: I would like to know why `upsert: true` is using here ?

File: src/app/(app)/attendance-report/components/storePdf.tsx
Comment (ghulamhafeez) [reply src/app/(app)/attendance-report/components/storePdf.tsx:1]: Fixed

File: src/app/(app)/attendance-report/components/AttendanceReport.tsx
Comment (ghulamhafeez) [reply src/app/(app)/attendance-report/components/AttendanceReport.tsx:1]: Fixed

File: src/app/(app)/attendance-report/components/generatePdf.tsx
Comment (ghulamhafeez) [reply src/app/(app)/attendance-report/components/generatePdf.tsx:1]: Fixed

File: tsconfig.json (hunk 1 above)
Comment (ghulamhafeez) [reply tsconfig.json:30]: Automatically generated , and this line is not necessary src/app/(app)/attendance-report/components/MyDocument.js

File: src/app/api/cron/notifications/monthly-attendance-report/route.tsx (hunk 2 above)
Comment (ghulamhafeez) [reply src/app/api/cron/notifications/monthly-attendance-report/route.tsx:42]: I used upsert to fix a duplicate error before, but now it's handled by a cron job, so I'll remove it.

File: src/app/api/cron/notifications/monthly-attendance-report/route.tsx [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import * as cron from "node-cron";
+import ReactPDF from "@react-pdf/renderer";
+import { NextResponse } from "next/server";
+import { getSupabaseClient } from "@/utils/supabase/supabaseClient";
+import { MonthlyAttendancePdf } from "@/app/(app)/attendance-report/components/MonthlyReportPdf";
+import { getAttendanceReport } from "@/app/(app)/attendance-report/actions/attendance-report";
+
+export async function GET() {
+  try {
+    cron.schedule("0 6 1 * *", async () => {
```
Comment (Monfernape) [inline src/app/api/cron/notifications/monthly-attendance-report/route.tsx:11]: Let's define constants for these cron string. They aren't very readable to everyone.

File: src/app/api/cron/notifications/monthly-attendance-report/route.tsx [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+          .on("data", (chunk) => chunks.push(chunk))
+          .on("end", () => resolve(Buffer.concat(chunks)))
+          .on("error", reject);
+      });
+    };
+
+    // eslint-disable-next-line react/react-in-jsx-scope
+    const stream = await ReactPDF.renderToStream(
+      // eslint-disable-next-line react/react-in-jsx-scope
+      <MonthlyAttendancePdf data={attendanceReport} />
```
Comment (Monfernape) [inline src/app/api/cron/notifications/monthly-attendance-report/route.tsx:36]: This should not be imported from main app directory. Keep it around in backend.

Comment (ghulamhafeez) [reply src/app/api/cron/notifications/monthly-attendance-report/route.tsx:36]: Fixed

File: src/app/api/cron/notifications/monthly-attendance-report/route.tsx (hunk 3 above)
Comment (ghulamhafeez) [reply src/app/api/cron/notifications/monthly-attendance-report/route.tsx:11]: Fixed

File: src/app/(app)/attendance-report/components/generatePdf.tsx [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        <Text style={styles.cell}>Name</Text>
+        <Text style={styles.cell}>Attendance %</Text>
+        <Text style={styles.cell}>Status</Text>
+        <Text style={styles.cell}>Present</Text>
+        <Text style={styles.cell}>Absent</Text>
+        <Text style={styles.cell}>Leave</Text>
+      </View>
+
+      {/* Table Rows */}
+      {data?.map((item, index) => (
```
[the remark this answers is not in the document]
Comment (ghulamhafeez) [reply src/app/(app)/attendance-report/components/generatePdf.tsx]: Fixed

File: src/app/api/attendance/route.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
+"use server";
+import React from "react";
+import { NextResponse } from "next/server";
+import ReactPDF from "@react-pdf/renderer";
+import { MyDocument } from "../../(app)/attendance-report/components/generatePdf";
+import path from "path";
+import { getAttendanceReport } from "@/app/(app)/attendance-report/actions/attendance-report";
+
+export async function handler(req) {
```
[the remark this answers is not in the document]
Comment (ghulamhafeez) [reply src/app/api/attendance/route.js]: This code is outdated

File: src/app/api/attendance/route.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+  const attendanceReport = await getAttendanceReport();
+  const fileName = `${new Date().toISOString().slice(0, 10)}.pdf`;
+  if (req.method === "GET") {
+    try {
+      const folderPath = path.join("Public/PDFs", "monthlyReport");
```
[the remark this answers is not in the document]
Comment (ghulamhafeez) [reply src/app/api/attendance/route.js]: This code is outdated

**Review by muhammad-bilal-ali-khokhar — APPROVED** [review]

**Pull request merged**